### ISSUE TITLE
Calling a struct with a manual implementation of `FnOnce` trips `clippy::no_effect` despite having side effects

### ISSUE THREAD

**Issue by PatchMixolydic** [rust-lang/rust-clippy#7792]
Lint name: `clippy::no_effect`

I tried this code ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=ef0cf18168440a92810107cf8bc078c2)):

```rust
#![feature(fn_traits, unboxed_closures)]

struct GreetStruct;

impl FnOnce<(&str,)> for GreetStruct {
    type Output = ();

    extern "rust-call" fn call_once(self, (who,): (&str,)) -> Self::Output {
        println!("hello {}", who);
    }
}

fn greet_fn(who: &str) {
    println!("hello {}", who);
}

fn main() {
    // no lint
    greet_fn("world");
    // no lint
    GreetStruct.call_once(("world",));
    // warning: statement with no effect
    GreetStruct("world");
}
```

I expected to see this happen: Clippy succeeds with no warnings.

Instead, this happened:
```
warning: statement with no effect
  --> src/main.rs:23:5
   |
23 |     GreetStruct("world");
   |     ^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(clippy::no_effect)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#no_effect
```

This statement does have an effect, as removing it changes the program's output from this:
```
hello world
hello world
hello world
```

to this:
```
hello world
hello world
```

### Meta

**Rust version (`rustc +nightly -Vv`):**
```
rustc 1.57.0-nightly (25ec82738 2021-10-05)
binary: rustc
commit-hash: 25ec8273855fde2d72ae877b397e054de5300e10
commit-date: 2021-10-05
host: x86_64-unknown-linux-gnu
release: 1.57.0-nightly
LLVM version: 13.0.0
```

**Comment by PatchMixolydic** [comment]
This lint isn't tripped for closures with compiler-generated implementations of `FnOnce`. This code produces no warnings ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=a20248006a73cac2ccbefe6ce76a3c37)):
```rust
fn x(f: impl FnOnce(&str)) {
    f("world");
}

fn main() {
    x(|who| println!("hello {}", who));
}
```

**Comment by PatchMixolydic** [comment]
This only seems to happen with unit structs, [making `GreetStruct` a tuple struct](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=d7dd00b920ef2d1d7b828f73e9ae5864) or <a aria-label="making GreetStruct a regular struct" href="https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=c7dfd8d8751b232f636fb7ff5d270253">a regular struct</a> seems to have no effect.
